Add documentation about OpenXR spatial entities (#11015)
godotengine/godot-docs: tutorials/xr/openxr_spatial_entities.rst

                _add_tracker(tracker)

    # A tracked managed by XRServer was changed.
    func _on_tracker_updated(_tracker_name: StringName, _type: int):
        # For now we ignore this, there aren't changes here we need to react
        # to and the instanced scene can react to this itself if needed.
        pass

    # A tracker was removed from our XRServer.
    func _on_tracker_removed(tracker_name: StringName, type: int):
        if type == XRServer.TRACKER_ANCHOR:
            var tracker : XRTracker = XRServer.get_tracker(tracker_name)
            if _managed_nodes.has(tracker):
                # We emit this right before we remove it!
                removed_spatial_entity.emit(_managed_nodes[tracker])

                # Remove the node.
                remove_child(_managed_nodes[tracker])

                # Queue free the node.
                _managed_nodes[tracker].queue_free()

                # And remove from our managed nodes.
                _managed_nodes.erase(tracker)

Spatial anchors
---------------

Spatial anchors allow us to map real world location in our virtual world in such a way that the

> XR runtime will keep track of these locations and adjust them as needed.

Reviewer: Use of singular "location" in "real world location" but use of plural "locations" in "... will keep track of these locations"

> If supported anchors can be made persistent which means the anchors will be recreated in the correct

Reviewer: This gets nitpicky since it's been a commonality in this PR to not include commas after dependent clauses but this is my first time noticing it.

Original comment was

> Add a comma after "If supported."

location when your application starts again.

You can think of use cases such as:
- placing virtual windows around your space that are recreated when your application restarts
- placing virtual objects on your table or on your walls and have them recreated

Spatial anchors are tracked using :ref:`OpenXRAnchorTracker<class_OpenXRAnchorTracker>` objects
registered with the XRServer.

When needed the location of the spatial anchor will be updated automatically, the pose on the
related tracker will be updated and thus the :ref:`XRAnchor3D<class_XRAnchor3D>` node will 

> reposition.

Reviewer: This is kinda nitpicky, but I would write it to add a comma after "When needed," and change the comma after "automatically" to a period to prevent this sentence from being a run on.

> When a spatial anchor has been made persistent a Universally Unique Identifier (or UUID) is

Reviewer: This gets nitpicky since it's been a commonality in this PR to not include commas after dependent clauses but this is my second(ish) time noticing it. 

Original comment was

> Comma after persistent as "When a spatial anchor has been made persistent" is a dependent clause.

Author: Like this? `When a spatial anchor has been made persistent, a Universally Unique Identifier (or UUID) is`

Reviewer: Yep, that would work.

persistent anchors.

.. note::
    If you remove the persistence of an anchor, the UUID is destroyed but the anchor is not
    removed automatically. You will need to react to the completion of making an anchor
    "unpersistent" and then clean up the anchor.
    Also you will get an error if you try to destroy an anchor that is still persistent.

To complete our anchor system we start with creating a scene that we'll set as the scene
to instantiate for anchors on our spatial manager node.

This scene should have an :ref:`XRAnchor3D<class_XRAnchor3D>` node as the root but nothing
else. We will add a script to it that will load a subscene that contains the actual visual
aspect of our anchor so we can create different anchors in our scene.
We'll assume the intention is to make these anchors persistent and save the path to this
subscene as meta data for our UUID.

.. code-block:: gdscript

    class_name OpenXRSpatialAnchor3D
    extends XRAnchor3D

    var anchor_tracker : OpenXRAnchorTracker
    var child_scene : Node
    var made_persistent : bool = false

    ## Return the scene path for our UUID

>     func get_scene_path(p_uuid: String) -> String:

Author: Still on the fence on this one, I religiously prefix parameters I create because of the likelihood class members already exist with this name and I wish Godot did as well.

        OpenXRSpatialAnchorCapability.remove_anchor(p_tracker)

    ## Retrieve the scene we've added for a given tracker (if any).
    func get_tracked_scene(p_tracker : XRTracker) -> XRNode3D:
        for node in get_children():
            if node is XRNode3D and node.tracker == p_tracker.name:
                return node

        return null

.. note::
    There seems to be a bit of magic going on in the code above.
    Whenever a spatial anchor is created or removed on our anchor capability,
    the related tracker object is created or destroyed.
    This results in the spatial manager adding or removing the child scene for this
    anchor. Hence we can rely on this here.

Plane tracking
--------------

Plane tracking allows us to detect surfaces such as walls, floors, ceilings and tables in
the players vicinity. This data could come from a room capture performed by the user at
any time in the past, or detected live by optical sensors.
The plane tracking extension doesn't make a distinction here.

.. note::
    Some XR runtimes do require vendor extensions to enable and/or configure this process
    but the data will be exposed through this extension.

> The code we write up above for the spatial manager will already detect our new planes.

Reviewer: "write" probably should be "wrote" for past tense.

    func _enter_tree():
        var openxr_interface : OpenXRInterface = XRServer.find_interface("OpenXR")
        if openxr_interface and openxr_interface.is_initialized():
            # Just in case our session hasn't started yet,
            # call our spatial context creation on start.
            openxr_interface.session_begun.connect(_setup_spatial_context)

            # And in case it is already up and running, call it already,
            # it will exit if we've called it too early.
            _setup_spatial_context()

    func _exit_tree():
        if spatial_context:
            # disconnect from our discovery signal.
            OpenXRSpatialEntityExtension.spatial_discovery_recommended.disconnect(_on_perform_discovery)

            # Free our spatial context, this will clean it up.
            OpenXRSpatialEntityExtension.free_spatial_context(spatial_context)
            spatial_context = RID()

        var openxr_interface : OpenXRInterface = XRServer.find_interface("OpenXR")
        if openxr_interface and openxr_interface.is_initialized():
            openxr_interface.session_begun.disconnect(_setup_spatial_context)

>     func _on_perform_discovery(p_spatial_context):

Author: Uhm, prime example of this breaking:
```
    var spatial_context : RID

    ...

    func _on_perform_discovery(p_spatial_context):
        # We get this signal for all spatial contexts, so exit if this is not for us
        if p_spatial_context != spatial_context:
            return
```

This is why I prefix my parameters!

    spatial entities that are no longer being tracked. This is only available with update snapshots.

    For capabilities that only rely on discovery snapshots you may wish to do a cleanup based on
    entities that are no longer part of the snapshot instead of relying on the state change.

Spatial entities
""""""""""""""""

With the above information we now know how to query our spatial entities and get information about
them but there is a little more we need to look at when it comes to the entities themselves.

In theory we're getting all our data from our snapshots however OpenXR has an extra API
where we create a spatial entity object from our entity id.
While this object exists the XR runtime knows that we are using this entity and that the
entity is not cleaned up early. This is a prerequisit for performing an update query on
this entity.

In our example code we do so by calling ``OpenXRSpatialEntityExtension.make_spatial_entity``.

Some spatial entity APIs will automatically create the object for us.
In this case we need to call ``OpenXRSpatialEntityExtension.add_spatial_entity`` to register
the created object with our implementation.

> Both functions return a RID that we can use in further functions that require our entity object.

Author: `an Resource ID`? not `a Resource ID`?

or anchors shared with us.

.. note::
    Sharing of anchors is currently not supported in the spatial entities specification.

As we shown in our example before we always start with creating a spatial context but now using the
:ref:`OpenXRSpatialCapabilityConfigurationAnchor<class_OpenXRSpatialCapabilityConfigurationAnchor>`
configuration object.
We'll show an example of this code after we discuss persistance scopes.
First we'll look at managing local anchors.

There is no difference in creating spatial anchors from what we've discussed around the built-in
logic. The only important thing is to pass your own spatial context as a parameter to
``OpenXRSpatialAnchorCapability.create_new_anchor``.

Making an anchor persistent requires you to wait until the anchor is tracking, this means that you
must perform update queries for any anchor you create so you can process state changes.

In order to enable making anchors persistent you also have to setup a persistence scope.
In the core of OpenXR two types of persistence scopes are supported:

.. list-table:: Persistence scopes
   :header-rows: 1

   * - Enum
     - Description
   * - OPENXR_SPATIAL_PERSISTENCE_SCOPE_SYSTEM_MANAGED

>      - Provides the application with read-only access (i.e. application cannot modify this store)

Author: Pretty sure I copied that out of OpenXRs documentation..

Eh yes I did:
<img width="976" height="407" alt="image" src="https://github.com/user-attachments/assets/bc898675-11df-41e6-9983-906e7b829f09" />

We'll need to upstream that fix :P (I well alter it here)

       to spatial entities persisted and managed by the system.
       The application can use the UUID in the persistence component for this store to correlate 
       entities across spatial contexts and device reboots.
   * - OPENXR_SPATIAL_PERSISTENCE_SCOPE_LOCAL_ANCHORS
     - Persistence operations and data access is limited to spatial anchors, on the same device,

>        for the same user and same app (using `make_anchor_persistent` and

Author: And here too, direct copy from the documentation :) I'll go for `same user and app`. :)

        var snapshot : RID = discovery_result.get_spatial_snapshot()
        if snapshot:
            # Process our snapshot result.
            _process_snapshot(snapshot, true)

            # And cleanup our snapshot
            OpenXRSpatialEntityExtension.free_spatial_snapshot(snapshot)

    func _process(_delta):
        if not spatial_context:
            return

        var entity_rids : Array[RID]
        for entity_id in entities:
            entity_rids.push_back(entities[entity_id].entity)

        # We just want our anchor component here.
        var snapshot : RID = OpenXRSpatialEntityExtension.update_spatial_entities(spatial_context, entity_rids, [ \
            OpenXRSpatialEntityExtension.OPENXR_SPATIAL_COMPONENT_TYPE_ANCHOR, \
        ])
        if snapshot:
            # Process our snapshot.
            _process_snapshot(snapshot)

            # And cleanup our snapshot.
            OpenXRSpatialEntityExtension.free_spatial_snapshot(snapshot)

    func _process_snapshot(p_snapshot, p_get_uuids):
        pass

Finally we can process our snapshot. Note that we are using :ref:`OpenXRAnchorTracker<class_OpenXRAnchorTracker>`

> as our tracker class as this already has all the support for anchors build in.

Reviewer: "build in" should be "built in" since it's past tense.

a given option:

.. list-table:: Marker tracking options
   :header-rows: 1

   * - Option
     - Check for support
     - Configuration object
   * - April tag
     - ``april_tag_is_supported``
     - :ref:`OpenXRSpatialCapabilityConfigurationAprilTag<class_OpenXRSpatialCapabilityConfigurationAprilTag>`
   * - Aruco
     - ``aruco_is_supported``
     - :ref:`OpenXRSpatialCapabilityConfigurationAruco<class_OpenXRSpatialCapabilityConfigurationAruco>`
   * - QR code
     - ``qrcode_is_supported``
     - :ref:`OpenXRSpatialCapabilityConfigurationQrCode<class_OpenXRSpatialCapabilityConfigurationQrCode>`
   * - Micro QR code
     - ``micro_qrcode_is_supported``
     - :ref:`OpenXRSpatialCapabilityConfigurationMicroQrCode<class_OpenXRSpatialCapabilityConfigurationMicroQrCode>`

Each option has its own configuration object that you can use when creating a spatial entity.

QR codes allow you to encode a string which is decoded by the XR runtime and accessible when a marker is found.
With April tags and Aruco markers binary data is encoded based which you again can access when a marker is found,
however you need to configure the detection with the correct decoding format.

> As an example we'll create a spatial context that will find QR codes adn Aruco markers.

Reviewer: Typo, "adn" should be "and."

                        entity_tracker.set_pose("default", center_pose, Vector3(), Vector3(), XRPose.XR_TRACKING_CONFIDENCE_HIGH)

                        entity_tracker.bounds_size = bounded2d_list.get_size(i)

                        entity_tracker.marker_type = marker_list.get_marker_type(i)
                        entity_tracker.marker_id = marker_list.get_marker_id(i)
                        entity_tracker.marker_data = marker_list.get_marker_data(p_snapshot, i)
                    else:
                        entity_tracker.invalidate_pose("default")

                    # We don't register our tracker until after we've set our initial data.
                    if register_with_xr_server:
                        XRServer.add_tracker(entity_tracker)

        # Any entities we've got left over, we can remove.
        for entity_id in org_entities:
            var entity_tracker : OpenXRMarkerTracker = entities[entity_id]
            entity_tracker.spatial_tracking_state = OpenXRSpatialEntityTracker.ENTITY_TRACKING_STATE_STOPPED
            XRServer.remove_tracker(entity_tracker)
            entities.erase(entity_id)

    **

>         Q : Should we be doing update queries here to get position changes for markers??

Author: Mostly a reminder to myself, I need to add this in. We now do this in the built-in logic as well.

.. _doc_openxr_spatial_entities:

OpenXR spatial entities
=======================

For any sort of augmented reality application you need to access real world information, and be able to
track real world locations. OpenXR's spatial entities API was introduced for this exact purpose.

It has a very modular design. The core of the API defines how real world entities are structured,
how they are found, and how information about them is stored and accessed.

> Various extensions are added on top that implement specific systems such as marker tracking,

Author: that to? Doesn't flow right.. maybe `Various extensions are added on top, which implement specific systems such as marker tracking,` ?

Vendors have the ability to implement and expose additional capabilities and component types that can be
used with the core API. For Godot these can be implemented in extensions. These implementations
however fall outside of the scope of this manual.

Finally it is important to note that the spatial entity system makes use of asynchronous functions.
This means that you can start a process, and then get informed of it finishing later on.

Setup
-----

In order to use spatial entities you need to enable the related project settings.
You can find these in the OpenXR section:

.. image:: img/openxr_spatial_entities_project_settings.webp

.. list-table:: Spatial entity settings
   :header-rows: 1

   * - Setting
     - Description
   * - Enabled
     - Enables the core of the spatial entities system. This must be enabled for any of the spatial
       entities systems to work.
   * - Enable spatial anchors
     - Enables the spatial anchors capability that allow creating and tracking spatial anchors.
   * - Enable persistent anchors
     - Enables the ability to make spatial anchors persistent. This means that their location is stored
       and can be retrieved in subsequent sessions.

>    * - Enabled built-in anchor detection

Reviewer 2: ```suggestion
   * - Enable built-in anchor detection
```

        XRServer.tracker_updated.connect(_on_tracker_updated)
        XRServer.tracker_removed.connect(_on_tracker_removed)

        # Set up existing trackers.
        var trackers : Dictionary = XRServer.get_trackers(XRServer.TRACKER_ANCHOR)
        for tracker_name in trackers:
            var tracker : XRTracker = trackers[tracker_name]
            if tracker and tracker is OpenXRSpatialEntityTracker:
                _add_tracker(tracker)

    # Exit tree is called whenever our node is removed from our scene.
    func _exit_tree():
        # Clean up our signals.
        XRServer.tracker_added.disconnect(_on_tracker_added)
        XRServer.tracker_updated.disconnect(_on_tracker_updated)
        XRServer.tracker_removed.disconnect(_on_tracker_removed)

        # Clean up trackers.
        for tracker in _managed_nodes:
            removed_spatial_entity.emit(_managed_nodes[tracker])
            remove_child(_managed_nodes[tracker])
            _managed_nodes[tracker].queue_free()

        _managed_nodes.clear()

    # See if this tracker should be managed by us and add it.
    func _add_tracker(tracker: OpenXRSpatialEntityTracker):

>         var new_node : XRAnchor3D

Reviewer 2: I think we use the style without a space before the colon in the docs:

```suggestion
        var new_node: XRAnchor3D
```

        if not new_node:
            # No scene defined or able to be instantiated? We're done!
            return

        # Set up and add to our scene.
        new_node.tracker = tracker.name
        new_node.pose = "default"
        _managed_nodes[tracker] = new_node
        add_child(new_node)

        added_spatial_entity.emit(new_node)

    # A new tracker was added to our XRServer.
    func _on_tracker_added(tracker_name: StringName, type: int):
        if type == XRServer.TRACKER_ANCHOR:
            var tracker : XRTracker = XRServer.get_tracker(tracker_name)
            if tracker and tracker is OpenXRSpatialEntityTracker:
                _add_tracker(tracker)

    # A tracked managed by XRServer was changed.
    func _on_tracker_updated(_tracker_name: StringName, _type: int):
        # For now we ignore this, there aren't any changes here we need to react
        # to and the instanced scene can react to this itself if needed.
        pass

    # A tracker was removed from our XRServer.
    func _on_tracker_removed(tracker_name: StringName, type: int):
        if type == XRServer.TRACKER_ANCHOR:

>             var tracker : XRTracker = XRServer.get_tracker(tracker_name)

Reviewer 2: ```suggestion
            var tracker: XRTracker = XRServer.get_tracker(tracker_name)
```

Spatial anchors are tracked using :ref:`OpenXRAnchorTracker<class_OpenXRAnchorTracker>` objects
registered with the XRServer.

When needed, the location of the spatial anchor will be updated automatically; the pose on the
related tracker will be updated and thus the :ref:`XRAnchor3D<class_XRAnchor3D>` node will 
reposition.

When a spatial anchor has been made persistent, a Universally Unique Identifier (or UUID) is
assigned to the anchor. You will need to store this with whatever information you need to
reconstruct the scene.
In our example code below we'll simply call ``set_scene_path`` and ``get_scene_path``, but you
will need to supply your own implementations for these functions.

In order to create a persistent anchor you need to follow a specific flow:
- Create the spatial anchor
- Wait until the tracking status changes to ``ENTITY_TRACKING_STATE_TRACKING``
- Make the anchor persistent
- Obtain the UUID and save it

When an existing persistent anchor is found a new tracker is added that has the UUID already
set. It is this difference in workflow that allows us to correctly react to new and existing
persistent anchors.

.. note::

>     If you remove the persistence of an anchor, the UUID is destroyed but the anchor is not

Reviewer 2: "remove the persistence of" sounds super awkward. Maybe:

```suggestion
    If you unpersist an anchor, the UUID is destroyed but the anchor is not
```

        if new_state == OpenXRSpatialEntityTracker.ENTITY_TRACKING_STATE_TRACKING and not made_persistent:
            # Only attempt to do this once.
            made_persistent = true

            # This warning is optional if you don't want to rely on persistence.
            if not OpenXRSpatialAnchorCapability.is_spatial_persistence_supported():
                push_warning("Persistent spatial anchors are not supported on this device!")
                return

            # Make this persistent, this will notify that the UUID changed on the anchor,
            # we can then store our scene path which we've already applied to our
            # tracked scene.
            OpenXRSpatialAnchorCapability.persist_anchor(anchor_tracker, RID(), Callable())

    func _on_uuid_changed() -> void:
        if anchor_tracker.uuid != "":
            made_persistent = true

            if child_scene:
                # If we already have a subscene, save that with the UUID.
                set_scene_path(anchor_tracker.uuid, child_scene.scene_file_path)
            else:
                # If we do not, look up the UUID in our stored cache.

>                 var scene_path : String = get_scene_path(anchor_tracker.uuid)

Reviewer 2: ```suggestion
                var scene_path: String = get_scene_path(anchor_tracker.uuid)
```

        # by our manager. We can thus continue assuming this has happened.

        var anchor_scene = get_tracked_scene(new_anchor)
        if not anchor_scene:
            push_error("Couldn't locate anchor scene for %s, has the manager been configured with an applicable anchor scene?" % [ new_anchor.name ])
            return
        if not anchor_scene is OpenXRSpatialAnchor3D:
            push_error("Anchor scene for %s is not an OpenXRSpatialAnchor3D scene, has the manager been configured with an applicable anchor scene?" % [ new_anchor.name ])
            return

        anchor_scene.set_child_scene(p_child_scene_path)

    ## Removes this spatial anchor from our scene.
    ## If the spatial anchor is persistent, the associated UUID will be cleared.
    func remove_spatial_anchor(p_anchor: XRAnchor3D):
        # Do we have anchor support?
        if not OpenXRSpatialAnchorCapability.is_spatial_anchor_supported():
            push_error("Spatial anchors are not supported on this device!")
            return

        var tracker : XRTracker = XRServer.get_tracker(p_anchor.tracker)
        if tracker and tracker is OpenXRAnchorTracker:

>             var anchor_tracker : OpenXRAnchorTracker = tracker

Reviewer 2: ```suggestion
            var anchor_tracker: OpenXRAnchorTracker = tracker
```

The root node for this scene must be an :ref:`XRAnchor3D<class_XRAnchor3D>` node.
We'll add a :ref:`StaticBody3D<class_StaticBody3D>` node as a child and add a
:ref:`CollisionShape3D<class_CollisionShape3D>` and :ref:`MeshInstance3D<class_MeshInstance3D>`
node as children of the static body.

.. image:: img/openxr_plane_anchor.webp

The static body and collision shape will allow us to make the plane interactable.

The mesh instance node allows us to apply a "hole punch" material to the plane,
when combined with passthrough this turns our plane into a visual occluder.
Alternatively we can assign a material that will visualize the plane for debugging.

We configure this material as the ``material_override`` material on our MeshInstance3D.
For our "hole punch" material, create a :ref:`ShaderMaterial<class_ShaderMaterial>`
and use the following code as the shader code:

.. code-block:: glsl

    shader_type spatial;
    render_mode unshaded, shadow_to_opacity;

    void fragment() {
        ALBEDO = vec3(0.0, 0.0, 0.0);
    }

We also need to add a script to our scene to ensure our collision and mesh are applied.

.. code-block:: gdscript

    extends XRAnchor3D

>     var plane_tracker : OpenXRPlaneTracker

Reviewer 2: ```suggestion
    var plane_tracker: OpenXRPlaneTracker
```

object. 
Of specific note is the ``plane_label`` property that, if available, identifies the type of surface.
Please consult the :ref:`OpenXRPlaneTracker<class_OpenXRPlaneTracker>` class documentation for
further information.

Marker tracking
---------------

Marker tracking detects specific markers in the real world. These are usually printed images such
as QR codes.

The API exposes support for 4 different codes, QR codes, Micro QR codes, Aruco codes, and April tags,
however XR runtimes are not required to support them all.

When markers are detected, :ref:`OpenXRMarkerTracker<class_OpenXRMarkerTracker>` objects are
instantiated and registered with the XRServer.

Our existing spatial manager code already detects these, all we need to do is create a scene
with an :ref:`XRAnchor3D<class_XRAnchor3D>` node at the root, save this, and assign it to the
spatial manager as the scene to instantiate for markers.

The marker tracker should be fully configured when assigned, so all that is needed is a
``_ready`` function that reacts to the marker data. Below is a template for the
required code:

.. code-block:: gdscript

    extends XRAnchor3D

>     var marker_tracker : OpenXRMarkerTracker

Reviewer 2: ```suggestion
    var marker_tracker: OpenXRMarkerTracker
```

                        pass
                OpenXRSpatialComponentMarkerList.MARKER_TYPE_MICRO_QRCODE:
                    var data = marker_tracker.get_marker_data()
                    if data.type_of() == TYPE_STRING:
                        # Data is a QR code as a string, usually a URL.
                        pass
                    elif data.type_of() == TYPE_PACKED_BYTE_ARRAY:
                        # Data is binary, can be anything.
                        pass
                OpenXRSpatialComponentMarkerList.MARKER_TYPE_ARUCO:
                    # Use marker_tracker.marker_id to identify the marker.
                    pass
                OpenXRSpatialComponentMarkerList.MARKER_TYPE_APRIL_TAG:
                    # Use marker_tracker.marker_id to identify the marker.
                    pass

As we can see, QR Codes provide a data block that is either a string or a byte array.
Aruco and April tags provide an ID that is read from the code.

It's up to your use case how best to link the marker data to the scene that needs to be loaded.
An example would be to encode the name of the asset you wish to display in a QR code.

> **Maybe see if we can work out at least one example here, need access to hardware that supports this, should have that soon**

Author: Owh, yeah that can be removed, the [spatial entities demo](https://github.com/BastiaanOlij/spatial-entities-demo) has a working example of marker tracking. You can see a full implementation of this code [here](https://github.com/BastiaanOlij/spatial-entities-demo/blob/master/spatial_entities/spatial_entities_marker.gd#L7)

I think in due time we should make a simpler example just for marker tracking (and a separate example for anchor and plane tracking) as part of our [Godot demo projects](https://github.com/godotengine/godot-demo-projects) and then just link to those from the documention.

Until then, the sample code presented here should be enough.

Backend access
--------------

For most purposes the core system, along with any vendor extensions, should be what most
users would use as provided.

> For those who are implementing vendor extensions, or those for who the built-in logic doesn't

Reviewer 2: ```suggestion
For those who are implementing vendor extensions, or those for whom the built-in logic doesn't
```

Spatial contexts
""""""""""""""""

A spatial context is the main object through which we query the spatial entities system.
Spatial contexts allow us to configure how we interact with one or more capabilities.

It's recommended to create a spatial context for each capability that you wish to interact
with, in fact, this is what Godot does for its built-in logic.

We start by setting the capability configuration objects for the capabilities we wish to
access.
Each capability will enable the components we support for that capability.
Settings can determine which components will be enabled.
We'll look at these configuration objects in more detail as we look at each supported capability.

Creating a spatial context is an asynchronous action. This means we ask the XR runtime to
create a spatial context, and at a point in the future the XR runtime will provide us
with the result.

The following script is the start of our example and can be added as a node to your scene.
It shows the creation of a spatial context for plane tracking,
and sets up our entity discovery.

.. code-block:: gdscript

    extends Node

>     var spatial_context : RID

Reviewer 2: I think I'm going to stop pointing these out. They can all be grep'd for and there's also the possibility that the docs style doesn't mandate this anyway :-)

```suggestion
    var spatial_context: RID
```

of what an update snapshot would look like for plane tracking if we needed one:

.. code-block:: gdscript

    ...

    func _process(_delta):
        if not spatial_context:
            return

        var entity_rids : Array[RID]
        for entity_id in entities:
            entity_rids.push_back(entities[entity_id].entity)

        var snapshot : RID = OpenXRSpatialEntityExtension.update_spatial_entities(spatial_context, entity_rids, [ \
                OpenXRSpatialEntityExtension.COMPONENT_TYPE_BOUNDED_2D, \
                OpenXRSpatialEntityExtension.COMPONENT_TYPE_PLANE_ALIGNMENT \
            ])
        if snapshot:
            # Process our snapshot.
            _process_snapshot(snapshot)

            # And clean up our snapshot.
            OpenXRSpatialEntityExtension.free_spatial_snapshot(snapshot)

Note that in our example here we're using the same ``_process_snapshot`` function to process the snapshot.
This makes sense in most situations. However if the components you've specified when creating the snapshot
are different between your discovery snapshot and your update snapshot, you have to take that you query

> different components into account.

Reviewer 2: I think some words got jumbled in this text: "you have to take that you query different components into account"
